Bug: Undo and card in learning.

Arthur-Milchior opened the pull request:
There are multiple things here.

I split undo in two, so that I get access to the actual background part without creating a background task nad having to deal with UI part. This is useful to entirely tests undo, since there are some parts of undo not done in col.undo.

Then, I add a test that bugs and should not bug. Thanks to David for explaining how to create it. Because of it, a card undone which becomes a learning card is not added to the learning queue. The problem is that `mCardNotToFetch` was not set to null even when it is not relevant anymore.

This was relevant for prefetching a card. Except that I forgot to tell the `preLoadCard` method to reset counts and queues if required, which made the method useless after a reset. (Actually, in the first version, preLoadCard was calling `getCard` so this would have been totally useless. But I did change for a version dealing with a queue of card and forgot to update this part...)

For safety, I also ensured that `getCard` set the two variables to null. This should be NF for reviewer. Maybe not NF for the content provider.

I should note it's not related to 2.12.1 bugs since those were introduce in alphas.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1125,6 +1126,33 @@ private TaskData doInBackgroundDismissNotes(TaskData param) {
     }
 
+    @VisibleForTesting
+    public static Card nonTaskUndo(Collection col) {
```
I can. But in my opinion, if there is a change to do, it would be `col.undo()` to actually do the entire part of the background task, dealing with the consequence of undoing such as telling the scheduler that queues and counts are not correct anymore.
Because, even if `col.undo()` is called from the content provider, that does not mean the scheduler has correct value, and currently the scheduler is not reset. 

I didn't change it because I was asked to avoid refactor that moves code and anyway I don't have a good intuition of who uses the content provider and whether they already deal with this fact themselves (assuming anyone actually undo from the content provider)

Arthur-Milchior replied:
That would be a natural fit for https://github.com/ankidroid/Anki-Android/pull/6959 by the way. 

I really believe that it's uselessly complex that "undo" deal with a method in background task, a method in collection before finally reaching the Undoable object. Especially since the job of the collection method is only to remove the Undo element from the queue, and the actual undoing is split between the CollectionTask method and the Undoable method for no reason I can imagine (it's not like we actually need a part of the undo to occur before the stack is changed and another part of undo to occur after the stack is changed)

Arthur-Milchior commented:
I did a version with and a version without preload

However, I realized that I should have used `.deferReset` in order to simulate leaving/coming back in the reviewer. in this case, it's not a regression test as there is no bug at all. And even reverting https://github.com/ankidroid/Anki-Android/pull/6960 does not lead to a bug. So I believe it's not a correct simulation of David's bug anymore. Note sure what would be a correct simulation

mikehardy commented:
Needs a conflict removal then a re-look from @david-allison-1 I think, @Arthur-Milchior ? This one is flagged for 2.13, to explain why I'm bumping it :-)

Arthur-Milchior commented:
Clearly, this new version need proof reading.

I went back and simplified the change I made. Now, there are only two states in the scheduler: with a card in reviewer or without card in the reviewer. Limits, counts and queue construction always take the reviewer's card into account instead of sometime doing it and sometime not. This limit drastically the number of member in the sched class. And actually, dealing with decrementing the limits is a simple process.

david-allison-1 commented:
```
com.ichi2.anki.tests.ContentProviderTest > testQueryCardFromCertainDeck[0][test(AVD) - 4.1.2] FAILED 
	java.lang.NullPointerException
	at com.ichi2.anki.tests.ContentProviderTest.testQueryCardFromCertainDeck(ContentProviderTest.java:834)
com.ichi2.anki.tests.ContentProviderTest > testQueryNextCard[0][test(AVD) - 4.1.2] FAILED 
	java.lang.NullPointerException
	at com.ichi2.anki.tests.ContentProviderTest.testQueryNextCard(ContentProviderTest.java:798)
com.ichi2.anki.tests.ContentProviderTest > testQueryCardFromCertainDeck[1][test(AVD) - 4.1.2] FAILED 
	java.lang.NullPointerException
	at com.ichi2.anki.tests.ContentProviderTest.testQueryCardFromCertainDeck(ContentProviderTest.java:834)
com.ichi2.anki.tests.ContentProviderTest > testQueryNextCard[1][test(AVD) - 4.1.2] FAILED 
	java.lang.NullPointerException
	at com.ichi2.anki.tests.ContentProviderTest.testQueryNextCard(ContentProviderTest.java:798)
```

Arthur-Milchior commented:
I corrected a bigger problem in tests. `sched.reset()` is never called directly outside of the scheduler. The point of the change `deferReset`was that the scheduler is responsible to know when reset should actually be executed. I replaced all `sched.reset()` by col.reset`, upstream they are equivalent, and that's the correct way to deal with it with the scheduler change of logic

david-allison-1 commented:
If we're never meant to call sched.reset then the method call needs to raise a warning. We can't just go about breaking the public API of a class without warning the callers that the method is now unusable

Arthur-Milchior commented:
Whyc API are you refering to ? The Card Content Provider does not have a RESET action so it seems not to be a problem.
The scheduler's reset is now protected, so it literally can't be used by other classes outside the scehduler class. 


To be more specific, any class can tell the scheduler it needs to reset because its values are probably invalid. The difference that was introduced with deferReset is simply that the actual computation is postponed until the values are needed. Because the profiler showed that anki spent uselessly time recomputing multiple time the counts during resets while it did not even need to show the counts until a deck is selected.

david-allison-1 commented:
Self-assigning for review: I still haven't logically worked my way through the changes

Arthur-Milchior commented:
Don't hesitate to ask any question.
Hopefully, each commit independently should be easy to follow.
